source/common/quic/quic_transport_socket_factory.h

```diff
@@ -11,44 +11,91 @@
 namespace Envoy {
 namespace Quic {
 
+#define QUIC_TRANSPORT_SOCKET_FACTORY_STATS(COUNTER)                                               \
+  COUNTER(context_config_update_by_sds)                                                            \
+  COUNTER(upstream_context_secrets_not_ready)                                                      \
+  COUNTER(downstream_context_secrets_not_ready)
+
+struct QuicTransportSocketFactoryStats {
+  QUIC_TRANSPORT_SOCKET_FACTORY_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+namespace {
+
+QuicTransportSocketFactoryStats generateStats(Stats::Scope& store, const std::string& perspective) {
+  return {QUIC_TRANSPORT_SOCKET_FACTORY_STATS(
+      POOL_COUNTER_PREFIX(store, fmt::format("quic_{}_transport_socket_factory.", perspective)))};
+}
+
+} // namespace
+
 // Base class for QUIC transport socket factory.
 // Because QUIC stack handles all L4 data, there is no need of a real transport
 // socket for QUIC in current implementation. This factory doesn't provides a
 // transport socket, instead, its derived class provides TLS context config for
 // server and client.
-class QuicTransportSocketFactoryBase : public Network::TransportSocketFactory {
+class QuicTransportSocketFactoryBase : public Network::TransportSocketFactory,
+                                       protected Logger::Loggable<Logger::Id::quic> {
 public:
+  QuicTransportSocketFactoryBase(Stats::Scope& store, std::unique_ptr<Ssl::ContextConfig> config,
+                                 const std::string& perspective)
+      : stats_(generateStats(store, perspective)), context_config_(std::move(config)) {
+    context_config_->setSecretUpdateCallback([this]() {
+      // The callback also triggers updating |context_config_| with the new secret.
+      // TODO(danzh) Client transport socket factory may also need to update quic crypto.
+      stats_.context_config_update_by_sds_.inc();
+    });
+  }
+
   // Network::TransportSocketFactory
   Network::TransportSocketPtr
   createTransportSocket(Network::TransportSocketOptionsSharedPtr /*options*/) const override {
     NOT_REACHED_GCOVR_EXCL_LINE;
   }
   bool implementsSecureTransport() const override { return true; }
   bool usesProxyProtocolOptions() const override { return false; }
+
+protected:
+  const Ssl::ContextConfig& contextConfig() const { return *context_config_; }
+  QuicTransportSocketFactoryStats stats_;
+
+private:
+  std::unique_ptr<Ssl::ContextConfig> context_config_;
```

Comment:
I don't have the full picture regarding the client context. However, if you can put `context_config_` to derived  `QuicServerTransportSocketFactory` and `QuicClientTransportSocketFactory`, the dynamic cast can be eliminated.

Reply:
Actually I moved it from subclasses to the base class so that they can have the SDS update callback setup in the base class :)
Doing so is just to have the same behavior on both endpoints. Indeed it brings up the need of dynamic_cast in subclasses. I don't have strong preference for either way.

Reply:
After thinking about it more, I changed the ownership to subclasses as the client subclass may need to do extra work in the future. Thus avoiding the dynamic_cast. Thanks for pointing out the issue here!